### DISCUSSION TITLE
Renovate isn't detecting composer updates when full versions aren't specified

### DISCUSSION THREAD

**Discussion by deviantintegral** [renovatebot/renovate#18291] [Request Help] [answered]
We noticed that Drupal wasn't being updated on one of our sites. We've created an example repository at https://github.com/Lullabot/renovate-composer-group-test

While the repository is currently set to `"rangeStrategy": "bump"`, `"rangeStrategy": "replace"` has the same issue.

In summary, as is Renovate isn't detecting any Drupal updates with the following configuration:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base"
  ],
  "rangeStrategy": "bump"
}
```

If we change the `require` from:

```
        "drupal/core-composer-scaffold": "^9.4",
        "drupal/core-recommended": "^9.4"
```

to:

```
        "drupal/core-composer-scaffold": "^9.4.3",
        "drupal/core-recommended": "^9.4.3"
```

Renovate properly opens PRs to update to `9.4.8`.

Running `composer update` locally properly upgrades the three packages (`drupal/core-composer-scaffold`, `drupal/core`, `drupal/core-recommended`).

Any other ideas? This eventually solves itself when using bumping, but is a bit confusing when onboarding an existing repo that may only specify major or minor versions.

**Comment by rarkins** [comment]
Possibly it's because in Composer, 9.4 == 9.4.0?

Is there any particular reason why you have `"drupal/core-composer-scaffold": "^9.4",` instead of `"drupal/core-composer-scaffold": "^9.4.0",` considering that you are expecting Renovate to bump to patch releases anyway?

**Reply by deviantintegral** [reply]
> Possibly it's because in Composer, 9.4 == 9.4.0?

I don't think that's the case because we were locked to 9.4.3, not 9.4.0.

> Is there any particular reason why you have "drupal/core-composer-scaffold": "^9.4",

Yes - this is the default with how composer works unless you manually specify the whole version. See how it uses `^1.4` but locks `1.4.1`:

```
$ composer require doctrine/instantiator
Using version ^1.4 for doctrine/instantiator
./composer.json has been updated
Running composer update doctrine/instantiator
Loading composer repositories with package information
Updating dependencies
Lock file operations: 1 install, 0 updates, 0 removals
  - Locking doctrine/instantiator (1.4.1)
Writing lock file
Installing dependencies from lock file (including require-dev)
Package operations: 1 install, 0 updates, 0 removals
  - Installing doctrine/instantiator (1.4.1): Extracting archive
Package doctrine/reflection is abandoned, you should avoid using it. Use roave/better-reflection instead.
Package symfony/debug is abandoned, you should avoid using it. Use symfony/error-handler instead.
Generating autoload files
43 packages you are using are looking for funding.
Use the `composer fund` command to find out more!
```

This behaviour is also throughout the Drupal and wider PHP ecosystem, such as in the composer file for new Drupal projects with `composer create-project drupal/recommended-project drupal`.

https://github.com/drupal/recommended-project/blob/9.4.8/composer.json#L17-L22

As well, it's the default copy-paste text when requiring modules, such as at https://www.drupal.org/project/type_tray

<img width="688" alt="image" src="https://user-images.githubusercontent.com/255023/195594567-8196ac7a-117e-4b49-abf6-af670ca961de.png">

With bumping, we're being more specific about ranges in that we prevent downgrades. We can solve this manually by also running `composer bump` locally, but it was still confusing to be missing known updates with security fixes.

**Reply by rarkins** [reply, marked as answer]
Fix: https://github.com/renovatebot/renovate/pull/18332
